### PR TITLE
r-modules: R bioc updater revised

### PR DESCRIPTION
**Pull request by PhDyellow**
This PR supercedes #122408, because I couldn't get the editorconfig checkers to pass on that branch, and merge conflicts had occured since the PR was submitted.

BioConductor releases are tied to specific R releases.

This pull request allows the package update scripts to automatically
select the correct version of bioconductor to match the current 
version of R.

Also update language documentation.

###### Motivation for this change

I updated the r-module package set (#122409, but also see #121819 which was submitted first and also bumps the package set) and encountered a few issues.

Bioconductor needed to have a version set in generate-r-packages.R, and the version must be compatible with the version of R, but that was not mentioned in the documentation. I have automated the version matching so the correct version of bioconductor will be chosen depending on the version of R used to update the packages. The maintainer should not have to update generate-r-packages.R before bumping anymore, as long as they use the R version in master.

I also had issues with some packages in bioc-experiment failing to get a hash. Re-running the scripts seemed to work most of the time.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/r-modules/default.nix
Comment (SuperSandro2000) [inline pkgs/development/r-modules/default.nix:46]:
```suggestion
    mkUrls = {name, version, biocVersion}: [
      "mirror://bioc/${biocVersion}/bioc/src/contrib/${name}_${version}.tar.gz"
```

File: pkgs/development/r-modules/default.nix
Comment (SuperSandro2000) [inline pkgs/development/r-modules/default.nix:49]:
```suggestion
      "mirror://bioc/${biocVersion}/bioc/src/contrib/Archive/${name}/${name}_${version}.tar.gz"
      "mirror://bioc/${biocVersion}/bioc/src/contrib/Archive/${name}_${version}.tar.gz"
    ];
  };
```

**Comment by PhDyellow** [comment]
@peti, given your comments in https://github.com/NixOS/nixpkgs/pull/93883#pullrequestreview-570738189, is this PR blocked until we can find another maintainer for R in NixOS?

File: doc/languages-frameworks/r.section.md
Comment (jbedo) [inline doc/languages-frameworks/r.section.md:107]: Which OS/filesystem has the locking issues? This doesn't sound like it's POSIX compliant, once the process exits the descriptors should be closed. I don't think we should be working around this by delaying an arbitrary amount.

File: doc/languages-frameworks/r.section.md
Comment (jbedo) [inline doc/languages-frameworks/r.section.md:117]: This shouldn't be part of the official update instructions. The provided shell should be used and not circumvented for expediency.

File: pkgs/development/r-modules/generate-r-packages.R
Comment (jbedo) [inline pkgs/development/r-modules/generate-r-packages.R:75]: These changes are unrelated to the commit message. What issue/bug is being resolved here? Can be split into a separate commit addressing the bug.

**Review by jbedo — CHANGES_REQUESTED** [review]
This is a great change overall.

**Review by jbedo — CHANGES_REQUESTED** [review]
~This commit doesn't seem relevant anymore, just introduces some whitespace? Can be dropped from the patch set?~ Sorry, github usage fail.

File: doc/languages-frameworks/r.section.md
Comment (PhDyellow) [reply doc/languages-frameworks/r.section.md:117]: Agree. I'll remove it.

File: doc/languages-frameworks/r.section.md
Comment (PhDyellow) [reply doc/languages-frameworks/r.section.md:107]: I use NixOS in a virtualbox VM, and I have cloned nixpkgs into a virtualbox shared folder. The shared folder points to an underlying NTFS formatted drive.

If that combination is not worth supporting, then I can remove this.

File: pkgs/development/r-modules/generate-r-packages.R
Comment (PhDyellow) [reply pkgs/development/r-modules/generate-r-packages.R:75]: I'll have to check this one, it doesn't look right to me anymore.

File: doc/languages-frameworks/r.section.md
Comment (jbedo) [reply doc/languages-frameworks/r.section.md:107]: I think that explains it, it's common for shared/distributed filesystems to not be fully POSIX compliant and race conditions like this can arise. This is difficult to handle as even waiting 1s may be insufficient for some filesystems. I'd rather not include this.

Comment (PhDyellow) [reply doc/languages-frameworks/r.section.md:107]: Sure, I will remove it.

**Comment by PhDyellow** [comment]
Rebasing so I test with the latest version of R.

**Comment by PhDyellow** [comment]
@jbedo,  @bcdarwin, I have been attempting to retest the changes in this PR, particularly the last code comment made: https://github.com/NixOS/nixpkgs/pull/127790#discussion_r699067271

Should the `nix-shell generate-shell.nix` be made pure?

My version of NixOS is out of date, I haven't had a chance to update recently.  When `generate-r-packages.R` attempts to call `nix-hash` it fails because of a mismatch between GLIBC versions. 
```
nix-hash: /nix/store/sbbifs2ykc05inws26203h0xwcadnf0l-glibc-2.32-46/lib/libc.so.6: version `GLIBC_2.33' not found (required by /nix/store/di1fqpncmjbjgvghqr6y4sqsrdi2653k-gfortran-9.3.0-lib/lib/libstdc++.so.6)
```
If I add `-pure` and call `nix-shell -pure generate-shell.nix` then SSL calls fail due to missing certificates, not surprising since the shell only pulls in R.

The quick fix is for me to update to the latest nixos-unstable, but I am able to put in a bit more time to get the shell working anywhere. Would it make a difference for others?

**Comment by PhDyellow** [comment]
`nix-shell generate-shell.nix` also rebuilt firefox, which doesn't make sense to me. What tools do you use to look at the dependency trees?

**Comment by jbedo** [comment]
You can:
```
nix-instantiate generate-shell.nix | xargs nix-store -q --tree
```
There's also `--graph` that produces Graphviz dot files.

Firefox get's pulled in via openjdk, required by the base R package.  I
agree this is excessive and we should take a look at it, but maybe not
in this PR.

**Comment by jbedo** [comment]
Would this be solved by adding nix to the `nativeBuildInputs` of
generate-shell.nix?

File: pkgs/development/r-modules/generate-r-packages.R
Comment (PhDyellow) [reply pkgs/development/r-modules/generate-r-packages.R]: Reverted, only seemed to matter when internet connectivity failed during package generation, but then other issues occured, where SHA fields came out empty.

**Comment by PhDyellow** [comment]
I tested some of the changes that @jbedo flagged, and found that they were not helping. I think I added them at a time when I was trying to update the package set over a flakey internet connection. When a package couldn't be downloaded, R didn't crash, it just returned an SHA of NULL. 

I have removed the changes, because they were not an attempt to make the package update script robust to poor internet, they were just hacks to get it to "pass" while still failing.

For the same reasons, I have reverted all the language documentation changes.

I believe this is ready to merge. As a bonus from testing, I have a bumped package set ready to put into a PR.

**Comment by PhDyellow** [comment]
I won't put together a variant of the package bump script that is robust to flakey internet, I don't think enough people would benefit from it.

**Comment by PhDyellow** [comment]
The bumped package set had a few conflicts with master, no PR at this stage.

File: pkgs/development/r-modules/generate-shell.nix
Comment (collares) [inline pkgs/development/r-modules/generate-shell.nix:10]: The spacing here looks a bit funny. Two options which are consistent with the rest of nixpkgs are:

```
  buildInputs = [
    wget
    cacert
    nix
  ];
  ```
or
```
  buildInputs = [ wget cacert nix ];
```

Comment (PhDyellow) [reply pkgs/development/r-modules/generate-shell.nix]: Thank you for the feedback, fixed and pushed.

**Comment by PhDyellow** [comment]
Whitespace fixed

**Review by jbedo — CHANGES_REQUESTED** [review]
Everything looks great and it all tests fine for me (in pure shell). I suggest one final change to cleanup the git history a bit and squash everything down to two commits:

1. the improved BIOC release logic;
2. the fixes to enable pure shell

**Comment by jbedo** [comment]
> The bumped package set had a few conflicts with master, no PR at this stage.

I've been working on an experimental update bot that runs daily, so you can pull the *-packages.nix files out of my tree https://github.com/jbedo/nixpkgs/tree/R if you want to avoid all the downloading and hashing. I merged your BIOC improvements into that branch to make the auto-update more reliable and I look forward to having your changes in nixpkgs.

File: pkgs/development/r-modules/default.nix
Comment (PhDyellow) [reply pkgs/development/r-modules/default.nix:48]: Sure, I added them in when I saw the whitespace changes requested in another PR review. I can't recall exactly which.

I will spin the whitespace changes from default.nix off into a new PR, we can accept or close the PR from there.

**Comment by PhDyellow** [comment]
I'm excited to see an automated package update for R! How often will in land in nixpkgs? If it won't happen often, I'll probably just point my development environment at your repo.

**Comment by PhDyellow** [comment]
The downloading and hashing just finished anyway. #137751

File: pkgs/development/r-modules/generate-shell.nix
Comment (SuperSandro2000) [inline pkgs/development/r-modules/generate-shell.nix:11]: Aren't cacert and nix part of the stdenv?

Comment (PhDyellow) [reply pkgs/development/r-modules/generate-shell.nix:11]: I couldn't get `generate-R-packages.R` to run without `cacert` and `nix`. 

`nix-shell generate-shell.nix` had access to the system packages, so SSL calls worked, but I hadn't updated NixOS in  a little while, and there was a mismatch in GLIBC, so `nix-hash` would not run.

`nix-shell --pure generate-shell` couldn't access the system certificates, so network calls failed due to missing SSL certificates. I think `nix-hash` was missing too. I'll check, but it might be a little while, I recently gc'd my nix store.

Comment (PhDyellow) [reply pkgs/development/r-modules/generate-shell.nix:11]: Is there a better way to bring those packages into a nix-shell?

Comment (PhDyellow) [reply pkgs/development/r-modules/generate-shell.nix:11]: Looking back at my commit messages, I did test `nix-shell --pure generate-shell` without `nix` and calls to `nix-hash` failed, saying it couldn't be found.

File: pkgs/development/r-modules/default.nix
Comment (PhDyellow) [reply pkgs/development/r-modules/default.nix:48]: The whitespace changes were requested by @SuperSandro2000 earlier in this PR https://github.com/NixOS/nixpkgs/pull/127790#discussion_r692869575

@jbedo, do you have a strong opinion on the matter?

Comment (jbedo) [reply pkgs/development/r-modules/default.nix:48]: No, I'm happy to have them merged as it's an improvement. If you could make it a separate commit that would be neatest.

File: pkgs/development/r-modules/generate-shell.nix
Comment (jbedo) [reply pkgs/development/r-modules/generate-shell.nix:11]: > Aren't cacert and nix part of the stdenv?

They are not. I'm a little surprised wget doesn't link to cacert as a fallback though.

**Comment by jbedo** [comment]
> I'm excited to see an automated package update for R! How often will in land in nixpkgs? If it won't happen often, I'll probably just point my development environment at your repo.

It's still early days, I have some issues to sort out (CRAN deletes packages and this causes problems) and before any automated PRs happen will need to sort out hydra so all the packages are built. I personally would want at least monthly updates since CRAN moves pretty quickly. If we select some "core packages" then we could generate a PR whenever they change (and successfully build).

Feel free to use my branch, I intend to keep it updating daily.

File: pkgs/development/r-modules/generate-shell.nix
Comment (PhDyellow) [reply pkgs/development/r-modules/generate-shell.nix:11]: I cannot make the shell work without these changes, and @jbedo has confirmed that this is expected behaviour in Nix. Getting cacert and nix into stdenv is obviously outside the scope of this PR. Marking this as resolved.

File: pkgs/development/r-modules/default.nix
Comment (PhDyellow) [reply pkgs/development/r-modules/default.nix:48]: It is a separate commit now.

**Comment by PhDyellow** [comment]
> It's still early days, I have some issues to sort out (CRAN deletes packages and this causes problems) and before any automated PRs happen will need to sort out hydra so all the packages are built. I personally would want at least monthly updates since CRAN moves pretty quickly. If we select some "core packages" then we could generate a PR whenever they change (and successfully build).
> 
> Feel free to use my branch, I intend to keep it updating daily.

Yes, CRAN deletes packages, and that messes with `default.nix`, when a package was manually fixed but is now gone. The only way to automate the fixes would be to inspect the system requirements for a package, but they aren't enforced by CRAN.

Would "conserving" dropped packages be OK? Basically, keep the last version that we hashed even if it is removed from CRAN? I think that would be simple to implement in `generate-R-packages`.

**Comment by jbedo** [comment]
> Would "conserving" dropped packages be OK? Basically, keep the last version that we hashed even if it is removed from CRAN? I think that would be simple to implement in `generate-R-packages`.

I'm not in favour of keeping packages around if they're not available upstream anymore, and hydra shouldn't test them either. We could keep the entry and mark it somehow as deprecated (or use meta.broken?) so it doesn't build.

**Comment by PhDyellow** [comment]
> I'm not in favour of keeping packages around if they're not available upstream anymore, and hydra shouldn't test them either. We could keep the entry and mark it somehow as deprecated (or use meta.broken?) so it doesn't build.

Fair enough, keeping orphaned packages isn't a great plan. I think tracking removed packages in default.nix would be a lot trickier though, with a lot of parsing of `default.nix` to decide which packages need to be marked as broken and which packages can just be forgotten about because they never touched `default.nix`. We might as well just automatically find lines in `default.nix` and remove them using a very careful regex.

**Comment by PhDyellow** [comment]
And we'd have to watch out for false positives, where a package was removed from CRAN, then the author fixes the package and it goes back up. With a daily update and automatic removal of lines from `default.nix`, we now have to go in and re-add all the overrides.

It's not obvious what the right approach is. But its probably time to move the discussion to a new place if you are actively working on it.

**Pull request merged**
